File: src/ModuleSplitter.java
package com.envr.modulesplitter;

import java.util.Arrays;

/**
 * Java implementation of Module Splitting Theorem.
 * Theorem: M = L ⊕ N with α, β, σ, ρ iff
 * βα=0, βσ=1, ρσ=0, ρα=1, αρ+σβ=1
 */
public class ModuleSplitter {
    private final int lDim;
    private final int nDim;
    private final int mDim;
    
    public ModuleSplitter(int lDim, int nDim) {
        this.lDim = lDim;
        this.nDim = nDim;
        this.mDim = lDim + nDim;
    }
    
    public static class Matrix {
        private final double[][] data;
        private final int rows;
        private final int cols;
        
        public Matrix(int rows, int cols) {
            this.rows = rows;
            this.cols = cols;
            this.data = new double[rows][cols];
        }
        
        public Matrix multiply(Matrix other) {
            if (this.cols != other.rows) {
                throw new IllegalArgumentException("Matrix dimensions incompatible");
            }
            
            Matrix result = new Matrix(this.rows, other.cols);
            for (int i = 0; i < this.rows; i++) {
                for (int j = 0; j < other.cols; j++) {
                    double sum = 0.0;
                    for (int k = 0; k < this.cols; k++) {
                        sum += this.data[i][k] * other.data[k][j];
                    }
                    result.data[i][j] = sum;
                }
            }
            return result;
        }
        
        public Matrix add(Matrix other) {
            if (this.rows != other.rows || this.cols != other.cols) {
                throw new IllegalArgumentException("Matrix dimensions must match");
            }
            
            Matrix result = new Matrix(rows, cols);
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    result.data[i][j] = this.data[i][j] + other.data[i][j];
                }
            }
            return result;
        }
        
        public boolean isZero(double tolerance) {
            for (double[] row : data) {
                for (double val : row) {
                    if (Math.abs(val) > tolerance) return false;
                }
            }
            return true;
        }
        
        public boolean isIdentity(double tolerance) {
            if (rows != cols) return false;
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    double expected = (i == j) ? 1.0 : 0.0;
                    if (Math.abs(data[i][j] - expected) > tolerance) return false;
                }
            }
            return true;
        }
        
        public void setIdentity() {
            int minDim = Math.min(rows, cols);
            for (int i = 0; i < minDim; i++) {
                data[i][i] = 1.0;
            }
        }
        
        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (double[] row : data) {
                sb.append(Arrays.toString(row)).append("\n");
            }
            return sb.toString();
        }
    }
    
    public Matrix createAlpha() {
        Matrix alpha = new Matrix(mDim, lDim);
        for (int i = 0; i < lDim; i++) {
            alpha.data[i][i] = 1.0;
        }
        return alpha;
    }
    
    public Matrix createBeta() {
        Matrix beta = new Matrix(nDim, mDim);
        for (int i = 0; i < nDim; i++) {
            beta.data[i][lDim + i] = 1.0;
        }
        return beta;
    }
    
    public Matrix createSigma() {
        Matrix sigma = new Matrix(mDim, nDim);
        for (int i = 0; i < nDim; i++) {
            sigma.data[lDim + i][i] = 1.0;
        }
        return sigma;
    }
    
    public Matrix createRho() {
        Matrix rho = new Matrix(lDim, mDim);
        for (int i = 0; i < lDim; i++) {
            rho.data[i][i] = 1.0;
        }
        return rho;
    }
    
    public boolean verifyTheorem(Matrix alpha, Matrix beta, Matrix sigma, Matrix rho) {
        double tolerance = 1e-10;
        
        Matrix ba = beta.multiply(alpha);
        Matrix bs = beta.multiply(sigma);
        Matrix rs = rho.multiply(sigma);
        Matrix ra = rho.multiply(alpha);
        Matrix ar = alpha.multiply(rho);
        Matrix sb = sigma.multiply(beta);
        Matrix arPlusSb = ar.add(sb);
        
        return ba.isZero(tolerance) &&
               bs.isIdentity(tolerance) &&
               rs.isZero(tolerance) &&
               ra.isIdentity(tolerance) &&
               arPlusSb.isIdentity(tolerance);
    }
    
    public static void main(String[] args) {
        ModuleSplitter splitter = new ModuleSplitter(2, 3);
        
        Matrix alpha = splitter.createAlpha();
        Matrix beta = splitter.createBeta();
        Matrix sigma = splitter.createSigma();
        Matrix rho = splitter.createRho();
        
        System.out.println("Module Splitting Theorem - Java Implementation");
        System.out.println("Dimensions: L=" + splitter.lDim + 
                         ", N=" + splitter.nDim + 
                         ", M=" + splitter.mDim);
        
        if (splitter.verifyTheorem(alpha, beta, sigma, rho)) {
            System.out.println("✓ Theorem verified: M = L ⊕ N");
        } else {
            System.out.println("✗ Theorem conditions not satisfied");
        }
    }
}

File: src/module_splitter.py
#!/usr/bin/env python3
"""
Module Splitting Theorem Proof Implementation
Proof that M = L ⊕ N with α, β, σ, ρ iff:
βα=0, βσ=1, ρσ=0, ρα=1, αρ+σβ=1
"""

import numpy as np
from typing import Tuple, List

class ModuleSplitter:
    """Implementation of the module splitting theorem."""
    
    def __init__(self, l_dim: int, n_dim: int):
        """
        Initialize with dimensions for L and N.
        M will have dimension l_dim + n_dim.
        """
        self.l_dim = l_dim
        self.n_dim = n_dim
        self.m_dim = l_dim + n_dim
        
    def create_standard_maps(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """
        Create standard maps for α=i_L, β=π_N, σ=i_N, ρ=π_L.
        Returns: (alpha, beta, sigma, rho)
        """
        # α: L → M (injection)
        alpha = np.zeros((self.m_dim, self.l_dim))
        alpha[:self.l_dim, :] = np.eye(self.l_dim)
        
        # β: M → N (projection)
        beta = np.zeros((self.n_dim, self.m_dim))
        beta[:, self.l_dim:] = np.eye(self.n_dim)
        
        # σ: N → M (injection)
        sigma = np.zeros((self.m_dim, self.n_dim))
        sigma[self.l_dim:, :] = np.eye(self.n_dim)
        
        # ρ: M → L (projection)
        rho = np.zeros((self.l_dim, self.m_dim))
        rho[:, :self.l_dim] = np.eye(self.l_dim)
        
        return alpha, beta, sigma, rho
    
    def verify_theorem(self, alpha: np.ndarray, beta: np.ndarray, 
                      sigma: np.ndarray, rho: np.ndarray) -> bool:
        """
        Verify the five conditions of the theorem.
        """
        conditions = [
            np.allclose(beta @ alpha, 0),          # βα = 0
            np.allclose(beta @ sigma, np.eye(self.n_dim)),  # βσ = 1_N
            np.allclose(rho @ sigma, 0),           # ρσ = 0
            np.allclose(rho @ alpha, np.eye(self.l_dim)),   # ρα = 1_L
            np.allclose(alpha @ rho + sigma @ beta, np.eye(self.m_dim))  # αρ+σβ=1_M
        ]
        
        return all(conditions)
    
    def test_random_maps(self, num_tests: int = 100) -> Tuple[int, int]:
        """
        Test random maps to see if conditions imply direct sum.
        """
        pass_conditions = 0
        pass_decomposition = 0
        
        for _ in range(num_tests):
            # Generate random maps
            alpha = np.random.randn(self.m_dim, self.l_dim)
            beta = np.random.randn(self.n_dim, self.m_dim)
            sigma = np.random.randn(self.m_dim, self.n_dim)
            rho = np.random.randn(self.l_dim, self.m_dim)
            
            if self.verify_theorem(alpha, beta, sigma, rho):
                pass_conditions += 1
                # Check if M = Im(α) ⊕ Im(σ)
                # (implementation details omitted for brevity)
                
        return pass_conditions, pass_decomposition

if __name__ == "__main__":
    # Example usage
    splitter = ModuleSplitter(l_dim=2, n_dim=3)
    alpha, beta, sigma, rho = splitter.create_standard_maps()
    
    print("Standard maps created:")
    print(f"α shape: {alpha.shape}, β shape: {beta.shape}")
    print(f"σ shape: {sigma.shape}, ρ shape: {rho.shape}")
    
    if splitter.verify_theorem(alpha, beta, sigma, rho):
        print("✓ All theorem conditions satisfied")
    else:
        print("✗ Theorem conditions failed")
